In [2]:
# Import libraries used for data handling and machine learning

import pandas as pd
import numpy as np

# For splitting dataset
from sklearn.model_selection import train_test_split

# Machine learning model
from sklearn.linear_model import LogisticRegression

# Evaluation metrics for classification
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [3]:
# Load the diabetes dataset from an online source

url = "https://raw.githubusercontent.com/jbrownlee/Datasets/master/pima-indians-diabetes.data.csv"

columns = [
    "Pregnancies","Glucose","BloodPressure","SkinThickness",
    "Insulin","BMI","DiabetesPedigreeFunction","Age","Outcome"
]

df = pd.read_csv(url, names=columns)

# Display first 5 rows
df.head()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1
3,1,89,66,23,94,28.1,0.167,21,0
4,0,137,40,35,168,43.1,2.288,33,1


In [4]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Pregnancies,768.0,3.845052,3.369578,0.000,1.00000,3.0000,6.00000,17.00
Glucose,768.0,120.894531,31.972618,0.000,99.00000,117.0000,140.25000,199.00
BloodPressure,768.0,69.105469,19.355807,0.000,62.00000,72.0000,80.00000,122.00
SkinThickness,768.0,20.536458,15.952218,0.000,0.00000,23.0000,32.00000,99.00
Insulin,768.0,79.799479,115.244002,0.000,0.00000,30.5000,127.25000,846.00
BMI,768.0,31.992578,7.884160,0.000,27.30000,32.0000,36.60000,67.10
DiabetesPedigreeFunction,768.0,0.471876,0.331329,0.078,0.24375,0.3725,0.62625,2.42
Age,768.0,33.240885,11.760232,21.000,24.00000,29.0000,41.00000,81.00
Outcome,768.0,0.348958,0.476951,0.000,0.00000,0.0000,1.00000,1.00


In [6]:
df1 = df.copy()

In [8]:
df['Outcome'].value_counts()

Outcome
0    500
1    268
Name: count, dtype: int64

In [9]:
# Check number of rows and columns

df.shape

(768, 9)

In [10]:
# Check number of rows and columns

df.shape

(768, 9)

In [11]:
# Features (input variables)
X = df.drop("Outcome", axis=1)

# Target variable (what we want to predict)
y = df["Outcome"]

In [12]:
# First split: Training set (70%) and Temporary set (30%)

X_train, X_temp, y_train, y_temp = train_test_split(
    X, y,
    test_size=0.30,
    random_state=42
)

print("Training data shape:", X_train.shape)
print("Temporary data shape:", X_temp.shape)

Training data shape: (537, 8)
Temporary data shape: (231, 8)


In [13]:
# Split the temporary data into validation and test sets

X_val, X_test, y_val, y_test = train_test_split(
    X_temp, y_temp,
    test_size=0.50,
    random_state=42
)

print("Validation data shape:", X_val.shape)
print("Test data shape:", X_test.shape)

Validation data shape: (115, 8)
Test data shape: (116, 8)


In [14]:
# Combine features and target for saving

train_df = pd.concat([X_train, y_train], axis=1)
val_df = pd.concat([X_val, y_val], axis=1)
test_df = pd.concat([X_test, y_test], axis=1)

# Save datasets
train_df.to_csv("train_data.csv", index=False)
val_df.to_csv("validation_data.csv", index=False)
test_df.to_csv("test_data.csv", index=False)

print("Train, validation, and test datasets saved successfully.")

Train, validation, and test datasets saved successfully.


In [15]:
#training the model 

In [16]:
# Create the Logistic Regression model

model = LogisticRegression(max_iter=1000)

# Train the model using training data
model.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

In [17]:
# Predict using validation data

val_predictions = model.predict(X_val)

val_predictions[:10]

array([1, 1, 1, 1, 1, 1, 0, 0, 0, 0], dtype=int64)

In [18]:
# Evaluate how well the model performs on validation data

val_accuracy = accuracy_score(y_val, val_predictions)

print("Validation Accuracy:", val_accuracy)

print("\nConfusion Matrix:")
print(confusion_matrix(y_val, val_predictions))

print("\nClassification Report:")
print(classification_report(y_val, val_predictions))

Validation Accuracy: 0.7304347826086957

Confusion Matrix:
[[61 17]
 [14 23]]

Classification Report:
              precision    recall  f1-score   support

           0       0.81      0.78      0.80        78
           1       0.57      0.62      0.60        37

    accuracy                           0.73       115
   macro avg       0.69      0.70      0.70       115
weighted avg       0.74      0.73      0.73       115



In [19]:
# Final evaluation using test data

test_predictions = model.predict(X_test)

test_accuracy = accuracy_score(y_test, test_predictions)

print("Test Accuracy:", test_accuracy)

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, test_predictions))

print("\nClassification Report:")
print(classification_report(y_test, test_predictions))

Test Accuracy: 0.7413793103448276

Confusion Matrix:
[[59 14]
 [16 27]]

Classification Report:
              precision    recall  f1-score   support

           0       0.79      0.81      0.80        73
           1       0.66      0.63      0.64        43

    accuracy                           0.74       116
   macro avg       0.72      0.72      0.72       116
weighted avg       0.74      0.74      0.74       116



# Diabetes Classification Model Summary

## Dataset Overview

In this example, we used the **Pima Indians Diabetes Dataset** to build a machine learning model that predicts whether a patient is diabetic or not.

The dataset contains **768 patient records** and **9 columns**.

### Features used for prediction

- Pregnancies
- Glucose
- BloodPressure
- SkinThickness
- Insulin
- BMI
- DiabetesPedigreeFunction
- Age

### Target Variable

Outcome

- 0 = Non-diabetic
- 1 = Diabetic

This makes the problem a **binary classification problem**.

---

# Data Splitting

The dataset was divided into three parts to simulate a real machine learning workflow.

Training Set → Used to train the model  
Validation Set → Used to evaluate and tune the model during development  
Test Set → Used for final evaluation on unseen data  

Dataset sizes after splitting:

- Training data: 537 samples (70%)
- Validation data: 115 samples (15%)
- Test data: 116 samples (15%)

---

# Model Used

We trained a **Logistic Regression model**.

Logistic Regression is commonly used for **binary classification problems** where the goal is to predict one of two possible outcomes.

In this case, the model learns relationships between the medical features and whether a patient is diabetic.

---

# Validation Results

Validation Accuracy: **0.73 (73%)**

This means the model correctly predicted about **73% of patients** in the validation dataset.

### Validation Confusion Matrix

| Actual / Predicted | Non-diabetic | Diabetic |
|---|---|---|
| Non-diabetic | 61 | 17 |
| Diabetic | 14 | 23 |

Interpretation:

- 61 patients were correctly predicted as non-diabetic
- 23 patients were correctly predicted as diabetic
- 17 patients were incorrectly predicted as diabetic
- 14 diabetic patients were missed by the model

### Validation Classification Report

Class 0 (Non-diabetic)

Precision: 0.81  
Recall: 0.78  
F1 Score: 0.80  

Class 1 (Diabetic)

Precision: 0.57  
Recall: 0.62  
F1 Score: 0.60  

The model performs slightly better at predicting **non-diabetic patients**, which is expected because the dataset contains more non-diabetic samples.

---

# Test Results

Test Accuracy: **0.74 (74%)**

The test dataset was **never used during training**, so this result represents how the model performs on completely unseen data.

### Test Confusion Matrix

| Actual / Predicted | Non-diabetic | Diabetic |
|---|---|---|
| Non-diabetic | 59 | 14 |
| Diabetic | 16 | 27 |

Interpretation:

- 59 non-diabetic patients were correctly predicted
- 27 diabetic patients were correctly predicted
- 14 patients were falsely predicted as diabetic
- 16 diabetic patients were missed

---

# Key Observations

- The model achieved around **74% accuracy** on unseen data.
- Validation accuracy and test accuracy are **very close**, which suggests the model is **generalizing well**.
- The dataset is **slightly imbalanced**, so the model performs better for the majority class (non-diabetic).

---

# Conclusion

This notebook demonstrated a complete machine learning workflow:

1. Loading and exploring the dataset
2. Splitting data into training, validation, and test sets
3. Training a Logistic Regression model
4. Evaluating performance using accuracy, confusion matrix, and classification metrics
5. Testing the model on unseen data

The model achieved a reasonable performance and provides a good example of how classification models are built and evaluated in real-world machine learning pipelines.